### PR TITLE
Isolate codegen

### PR DESCRIPTION
**Pull request by adamwight**

### CODE REVIEW

**Comment by adamwight** [comment]
From halfak in IRC,
```
awight, seems like main_template should be the template itself rather than a file to look-up.  That'd make it easier to write a test for.
10:07 AM You can then attempt to load the template in the main() function -- which is a far nicer place to throw an error like FileNotFound. 
10:08 AM You can have a nice function in codegen called "read_template" or something. 
10:08 AM that uses default paths or whatever. 
10:09 AM Maybe the default path is built into generate_make
```

It's tricky because we're loading the main template through the Jinja environment and loader, which shouldn't be referenced from outside of the codegen package.  The Jinja loader will throw a TemplateNotFound exception, so I don't think there's a benefit in forcing the FileNotFound instead.

**Comment by halfak** [comment]
Here's my first search result: http://matthiaseisen.com/pp/patterns/p0198/
Seems like we could do something similar.

Alternatively, we could have it take the template as a string.  That would be the easiest for testing:

```
from jinja2 import Environment, BaseLoader

rtemplate = Environment(loader=BaseLoader).from_string(myString)
```

File: editquality/codegen/generate.py [hunk 1]
```diff
@@ -1,19 +1,24 @@
 import jinja2
 
-from . import config
 
+def generate(variables, templates_path, main_template):
```
Comment (halfak) [inline editquality/codegen/generate.py:4]: templates_path seems to be unnecessary

File: editquality/codegen/generate.py [hunk 2]
```diff
@@ -1,19 +1,24 @@
...
-
-    # TODO: map config into domain-specific preprocessed values
+    Returns the rendered output.
+    """
 
     env = jinja2.Environment(
         loader=jinja2.FileSystemLoader(templates_path),
         lstrip_blocks=True,
         trim_blocks=True
     )
```
Comment (halfak) [inline editquality/codegen/generate.py:21]: Could use BaseLoader

Comment (adamwight) [reply editquality/codegen/generate.py:21]: I'm planning to introduce partials, which require a filesystem lookup...

Comment (halfak) [reply editquality/codegen/generate.py:21]: What do you mean by partials?  Is this like including one j2 template in another?  If so, I think I see where you are coming from.

Comment (adamwight) [reply editquality/codegen/generate.py:21]: Exactly.  Here's the documentation, http://jinja.pocoo.org/docs/2.10/templates/#include

I'm still fine with making the testability change and passing string template contents, cos I like that the calling function has more control.  Maybe we should allow overriding the loader, in the future.

Comment (halfak) [reply editquality/codegen/generate.py:21]: OK this makes sense to me.

**Pull request merged**